File: 127-word-ladder/word-ladder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {
    const wordSet = new Set(wordList);
    const queue = [];

    if(!wordSet.has(endWord)) {
        return 0;
    }

    queue.push([beginWord, 1]);

    while(queue.length) {
        const [currWord, level] = queue.shift();
        for(let i=0; i<currWord.length; i++) {
            for(let j=0; j<26; j++) {
                const nextChar = String.fromCharCode("a".charCodeAt(0) + j);
                if(nextChar !== currWord[i]) {
                    const newWord = currWord.slice(0, i) + nextChar + currWord.slice(i+1);
                    if(newWord === endWord) {
                        return level + 1;
                    }
                    if(wordSet.has(newWord)) {
                        queue.push([newWord, level + 1]);
                        wordSet.delete(newWord);
                    }
                }
            }
        }
    }

    return 0;
};

File: 127-word-ladder/word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord,
                     vector<string>& wordList) {
        queue<string> q;
        q.push(beginWord);

        unordered_set<string> wordSet(
            wordList.begin(),
            wordList.end()); // for faster search of words + make sure visited
                             // words don't appear again

        // endWord should be in the list in first place
        if(!wordSet.count(endWord)) return 0; 

        int level = 1;
        while (!q.empty()) {
            int size = q.size();
            while (size--) {
                string word = q.front();
                q.pop();

                if (word == endWord)
                    return level; // FOUND

                // hot(*ot, h*t, ho*) -> in place of * try a-z and search for
                // possible nbrs in wordSet
                for (int i = 0; i < word.size(); i++) {
                    string nbr = word;
                    for (char ch = 'a'; ch <= 'z'; ch++) {
                        nbr[i] = ch; // replacing *(at pos i) with a-z

                        if (wordSet.count(nbr)) {
                            q.push(nbr);
                            wordSet.erase(nbr); // Like marking nbr as visited
                        }
                    }
                }
            }
            level++;
        }

        return 0; // End word not found
    }
};